File: wpilibj/src/main/java/org/wpilib/simulation/GearboxSim.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package org.wpilib.simulation;

import org.wpilib.math.linalg.VecBuilder;
import org.wpilib.math.numbers.N1;
import org.wpilib.math.numbers.N2;
import org.wpilib.math.system.Gearbox;
import org.wpilib.math.system.LinearSystem;
import org.wpilib.system.RobotController;

/**
 * Represents a simulated gearbox driving a rotational mechanism.
 *
 * <p>Positions, velocities and torques are those of the gearbox's output shaft; current draw is the
 * total across all of its motors.
 */
public class GearboxSim extends LinearSystemSim<N2, N1, N2> {
  // Gearbox for the sim.
  private final Gearbox m_gearbox;

  // The moment of inertia of the driven mechanism in kg-m².
  private final double m_j;

  /**
   * Creates a simulated gearbox.
   *
   * @param plant The linear system representing the mechanism. This system can be created with
   *     {@link org.wpilib.math.system.Models#singleJointedArmFromPhysicalConstants(Gearbox,
   *     double)} or {@link org.wpilib.math.system.Models#singleJointedArmFromSysId(double,
   *     double)}. It must have been built with the same reduction as {@code gearbox}.
   * @param gearbox The gearbox driving the mechanism.
   * @param measurementStdDevs The standard deviations of the measurements. Can be omitted if no
   *     noise is desired. If present must have 2 elements. The first element is for position. The
   *     second element is for velocity.
   */
  public GearboxSim(LinearSystem<N2, N1, N2> plant, Gearbox gearbox, double... measurementStdDevs) {
    super(plant, measurementStdDevs);
    m_gearbox = gearbox;

    // By theorem 6.10.1 of https://file.tavsys.net/control/controls-engineering-in-frc.pdf,
    // the DC motor state-space model is:
    //
    //   dx/dt = -G²Kₜ/(KᵥRJ)x + (GKₜ)/(RJ)u
    //   A = -G²Kₜ/(KᵥRJ)
    //   B = GKₜ/(RJ)
    //
    // Solve for J.
    //
    //   B = GKₜ/(RJ)
    //   J = GKₜ/(RB)
    //
    // Kₜ here is that of the gearbox as a whole, so it scales with the number of
    // motors. G is taken from the gearbox, so the plant must have been built with
    // the same reduction.
    m_j =
        gearbox.reduction
            * gearbox.numMotors
            * gearbox.motor.Kt
            / (gearbox.motor.R * plant.getB(1, 0));
  }

  /**
   * Sets the state of the gearbox.
   *
   * @param angularPosition The new position in radians.
   * @param angularVelocity The new velocity in radians per second.
   */
  public void setState(double angularPosition, double angularVelocity) {
    setState(VecBuilder.fill(angularPosition, angularVelocity));
  }

  /**
   * Sets the gearbox's angular position.
   *
   * @param angularPosition The new position in radians.
   */
  public void setAngle(double angularPosition) {
    setState(angularPosition, getAngularVelocity());
  }

  /**
   * Sets the gearbox's angular velocity.
   *
   * @param angularVelocity The new velocity in radians per second.
   */
  public void setAngularVelocity(double angularVelocity) {
    setState(getAngularPosition(), angularVelocity);
  }

  /**
   * Returns the moment of inertia of the driven mechanism.
   *
   * @return The mechanism's moment of inertia in kg-m².
   */
  public double getJ() {
    return m_j;
  }

  /**
   * Returns the gearbox.
   *
   * @return The gearbox.
   */
  public Gearbox getGearbox() {
    return m_gearbox;
  }

  /**
   * Returns the gearbox's angular position.
   *
   * @return The gearbox's position in radians.
   */
  public double getAngularPosition() {
    return getOutput(0);
  }

  /**
   * Returns the gearbox's angular velocity.
   *
   * @return The gearbox's velocity in radians per second.
   */
  public double getAngularVelocity() {
    return getOutput(1);
  }

  /**
   * Returns the gearbox's angular acceleration.
   *
   * @return The gearbox's acceleration in rad/s².
   */
  public double getAngularAcceleration() {
    var acceleration = (m_plant.getA().times(m_x)).plus(m_plant.getB().times(m_u));
    return acceleration.get(1, 0);
  }

  /**
   * Returns the torque at the gearbox's output.
   *
   * @return The output torque in Newton-meters.
   */
  public double getTorque() {
    return getAngularAcceleration() * m_j;
  }

  /**
   * Returns the total current drawn by the gearbox's motors.
   *
   * @return The gearbox's current draw in amps.
   */
  public double getCurrentDraw() {
    // Gearbox.getCurrent() takes the velocity of the output and applies the reduction
    // internally to get the motor velocity, and returns the total across all the motors.
    return m_gearbox.getCurrent(m_x.get(1, 0), m_u.get(0, 0)) * Math.signum(m_u.get(0, 0));
  }

  /**
   * Gets the input voltage for the gearbox.
   *
   * @return The gearbox's input voltage.
   */
  public double getInputVoltage() {
    return getInput(0);
  }

  /**
   * Sets the input voltage for the gearbox.
   *
   * @param volts The input voltage.
   */
  public void setInputVoltage(double volts) {
    setInput(volts);
    clampInput(RobotController.getBatteryVoltage());
  }
}

File: wpilibc/src/main/native/cpp/simulation/GearboxSim.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "wpi/simulation/GearboxSim.hpp"

#include "wpi/system/RobotController.hpp"
#include "wpi/util/MathExtras.hpp"

using namespace wpi;
using namespace wpi::sim;

GearboxSim::GearboxSim(const wpi::math::LinearSystem<2, 1, 2>& plant,
                       const wpi::math::Gearbox& gearbox,
                       const std::array<double, 2>& measurementStdDevs)
    : LinearSystemSim<2, 1, 2>(plant, measurementStdDevs),
      m_gearbox(gearbox),
      // By theorem 6.10.1 of
      // https://file.tavsys.net/control/controls-engineering-in-frc.pdf, the
      // DC motor state-space model is:
      //
      //   dx/dt = -G²Kₜ/(KᵥRJ)x + (GKₜ)/(RJ)u
      //   A = -G²Kₜ/(KᵥRJ)
      //   B = GKₜ/(RJ)
      //
      // Solve for J.
      //
      //   B = GKₜ/(RJ)
      //   J = GKₜ/(RB)
      //
      // Kₜ here is that of the gearbox as a whole, so it scales with the number
      // of motors. G is taken from the gearbox, so the plant must have been
      // built with the same reduction.
      m_j(gearbox.reduction * gearbox.numMotors * gearbox.motor.Kt.value() /
          (gearbox.motor.R.value() * m_plant.B(1, 0))) {}

void GearboxSim::SetState(wpi::units::radian_t angularPosition,
                          wpi::units::radians_per_second_t angularVelocity) {
  SetState(wpi::math::Vectord<2>{angularPosition, angularVelocity});
}

void GearboxSim::SetAngle(wpi::units::radian_t angularPosition) {
  SetState(angularPosition, GetAngularVelocity());
}

void GearboxSim::SetAngularVelocity(
    wpi::units::radians_per_second_t angularVelocity) {
  SetState(GetAngularPosition(), angularVelocity);
}

wpi::units::radian_t GearboxSim::GetAngularPosition() const {
  return wpi::units::radian_t{GetOutput(0)};
}

wpi::units::radians_per_second_t GearboxSim::GetAngularVelocity() const {
  return wpi::units::radians_per_second_t{GetOutput(1)};
}

wpi::units::radians_per_second_squared_t GearboxSim::GetAngularAcceleration()
    const {
  return wpi::units::radians_per_second_squared_t{
      (m_plant.A() * m_x + m_plant.B() * m_u)(1, 0)};
}

wpi::units::newton_meter_t GearboxSim::GetTorque() const {
  return wpi::units::newton_meter_t{GetAngularAcceleration().value() *
                                    m_j.value()};
}

wpi::units::ampere_t GearboxSim::GetCurrentDraw() const {
  // Gearbox::Current() takes the velocity of the output and applies the
  // reduction internally to get the motor velocity, and returns the total
  // across all the motors.
  return m_gearbox.Current(wpi::units::radians_per_second_t{m_x(1)},
                           wpi::units::volt_t{m_u(0)}) *
         wpi::util::sgn(m_u(0));
}

wpi::units::volt_t GearboxSim::GetInputVoltage() const {
  return wpi::units::volt_t{GetInput(0)};
}

void GearboxSim::SetInputVoltage(wpi::units::volt_t voltage) {
  SetInput(wpi::math::Vectord<1>{voltage.value()});
  ClampInput(wpi::RobotController::GetBatteryVoltage().value());
}

const wpi::math::Gearbox& GearboxSim::GetGearbox() const {
  return m_gearbox;
}

wpi::units::kilogram_square_meter_t GearboxSim::GetJ() const {
  return m_j;
}
